File: rmq.py
from sys import stdin
from math import log

class RMQ:
    def __init__(self, arr):
        n, k = len(arr), int(log(len(arr))/log(2))+1
        self.sparse = [[0 for _ in range(k)] for _ in range(n)]
        for i in range(n): self.sparse[i][0] = arr[i]

        j = 1
        while 1<<j <= n:
            i = 0
            while (i + (1 << j) - 1) < n:
                if self.sparse[i][j-1] < self.sparse[i+(1<<(j-1))][j-1]:
                    self.sparse[i][j] = self.sparse[i][j-1]
                else:
                    self.sparse[i][j] = self.sparse[i+(1<<(j-1))][j-1]
                i += 1
            j+=1

    def query(self, L, R):
        j = int(log(R-L+1)/log(2))
        if self.sparse[L][j] <= self.sparse[R-(1<<j) + 1][j]: return self.sparse[L][j]
        else: return self.sparse[R - (1 << j) + 1][j];


def main():
    pass

main()

File: rmq.cpp
#include <bits/stdc++.h>
using namespace std;

class RMQ{
public:
    vector<vector<int>> sparse;
    RMQ(vector<int> &arr){
        int n = arr.size(), k = ((int)log2(arr.size()))+1;
        for(int i = 0 ; i < n ; i++) sparse.push_back(vector<int>(k,0));

        for (int i = 0; i < n; i++) sparse[i][0] = arr[i];

        for (int j = 1; (1 << j) <= n; j++) {
            for (int i = 0; (i + (1 << j) - 1) < n; i++) {
                if (sparse[i][j - 1] < sparse[i + (1 << (j - 1))][j - 1])
                    sparse[i][j] = sparse[i][j - 1];
                else
                    sparse[i][j] = sparse[i + (1 << (j - 1))][j - 1];
            }
        }
    }

    int query(int L, int R){
        int j = (int)log2(R - L + 1);
        if (sparse[L][j] <= sparse[R - (1 << j) + 1][j]) return sparse[L][j];
        else return sparse[R - (1 << j) + 1][j];
    }
};


int main()
{
    vector<int> a = { 7, 2, 3, 0, 5, 10, 3, 18, 12 };
    RMQ rmq = RMQ(a);
    cout << rmq.query(0, 4) << endl;
    cout << rmq.query(4, 7) << endl;
    cout << rmq.query(7, 8) << endl;
    return 0;
}
